db/user_test.go

```diff
@@ -14,6 +15,88 @@ type SyncNewUsersTestCase struct {
     sendEmails bool
 }
 
+func (this *DBTests) DBTestResolveUsers(test *testing.T) {
+    defer ResetForTesting();
+
+    testCases := []struct {input []string; expectedOutput []string; addUsers []*model.User; removeUsers []string} {
+        // This is a simple test case for the empty string input.
+        {
+            []string{""},
+            []string{},
+            nil,
+            []string{},
+        },
+
+        // This is a test to ensure the output is sorted.
+        {
+            []string{"b@test.com", "a@test.com", "c@test.com"},
+            []string{"a@test.com", "b@test.com", "c@test.com"},
+            nil,
+            []string{},
+        },
+
+        // This is a basic test to ensure that a role gets mapped to the correct email.
+        {
+            []string{"admin"},
+            []string{"admin@test.com"},
+            nil,
+            []string{},
+        },
+
+        // This is a test for our all roles character, the *. 
+        {
+            []string{"*"},
+            []string{"admin@test.com", "grader@test.com", "other@test.com", "owner@test.com", "student@test.com"},
+            nil,
+            []string{},
+        },
+
+        // This is a more complex test that adds multiple Users for the student role.
+        // This test case removes the only user from the "other" role, so we check that a role without any users still functions properly.
```

Comment:
Try to only do one thing per test case (unless that case is specifically for the interaction of different components).
Like here, do the duplicate users, add user, and caps all in different test cases.